### PR TITLE
Fix/rectangle to polygon

### LINKED ISSUE #1604: Rectangle --> Quadrilateral
Our rectangle isn't constrained and should be called a Quadrilateral.

### PR DESCRIPTION
**Pull request by hippo91**
### Overview

This PR address the problem of having a Rectangle function that behaves as a Quadrilateral factory.

resolves #1604

### Details

- The Rectangle function has been renamed in Quadrilateral.
- A new Rectangle function has been created. 
- It checks that only three points are given. If four points are given it raises a deprecation warning else if more or less than 3 points are given a TypeError is raised.
- The 3 points can be given in arbitrary order but should be different and defines orthogonal vectors otherwise ValueError is raised
- Extensive tests have been added.

### CODE REVIEW

File: pyvista/utilities/geometric_objects.py [hunk 1]
```diff
@@ -1253,9 +1257,85 @@ def Rectangle(points=None):
...
+        points[3] = point_3
+        cells = np.array([[4, 0, 1, 3, 2]])
+    if np.isclose(scalar_pdct_01_12, 0):
+        point_3 = point_1 + vec_12 - vec_01
+        points[3] = point_3
+        cells = np.array([[4, 0, 1, 2, 3]])
+    if np.isclose(scalar_pdct_02_12, 0):
+        point_3 = point_2 - vec_02 - vec_12
+        points[3] = point_3
+        cells = np.array([[4, 0, 2, 1, 3]])
```
Comment (tkoyama010) [inline pyvista/utilities/geometric_objects.py:1312]:
```suggestion
    if np.isclose(scalar_pdct_01_02, 0):
        points[3] = point_0 + vec_01 + vec_02
        cells = np.array([[4, 0, 1, 3, 2]])
    if np.isclose(scalar_pdct_01_12, 0):
        points[3] = point_1 + vec_12 - vec_01
        cells = np.array([[4, 0, 1, 2, 3]])
    if np.isclose(scalar_pdct_02_12, 0):
        points[3] = point_2 - vec_02 - vec_12
        cells = np.array([[4, 0, 2, 1, 3]])
```

File: pyvista/utilities/geometric_objects.py [hunk 2]
```diff
@@ -1264,9 +1268,82 @@ def Rectangle(points=None):
...
+        points[3] = point_0 + vec_01 + vec_02
+        cells = np.array([[4, 0, 1, 3, 2]])
+    if np.isclose(scalar_pdct_01_12, 0):
+        points[3] = point_1 + vec_12 - vec_01
+        cells = np.array([[4, 0, 1, 2, 3]])
+    if np.isclose(scalar_pdct_02_12, 0):
+        points[3] = point_2 - vec_02 - vec_12
+        cells = np.array([[4, 0, 2, 1, 3]])
+
+    return pyvista.wrap(pyvista.PolyData(points, cells))
```
Comment (adeak) [inline pyvista/utilities/geometric_objects.py:1322]: Why are we wrapping PolyData?

Comment (hippo91) [reply pyvista/utilities/geometric_objects.py:1322]: In fact i just took what existed in the former Rectangle function. This also what is done in the other function such a Triangle, Circle and Ellipse.

Comment (tkoyama010) [reply pyvista/utilities/geometric_objects.py:1322]: This seems to be an existing redundant code. I decided to remove the other parts with https://github.com/pyvista/pyvista/pull/4146.
```suggestion
    return pyvista.PolyData(points, cells)
```

File: pyvista/utilities/geometric_objects.py [hunk 3]
```diff
@@ -1264,9 +1268,82 @@ def Rectangle(points=None):
...
+
+    vec_01 = point_1 - point_0
+    vec_02 = point_2 - point_0
+    vec_12 = point_2 - point_1
+
+    scalar_pdct_01_02 = np.dot(vec_01, vec_02)
+    scalar_pdct_01_12 = np.dot(vec_01, vec_12)
+    scalar_pdct_02_12 = np.dot(vec_02, vec_12)
+
+    non_null_scalar_products = [
```
Comment (MatthewFlamm) [inline pyvista/utilities/geometric_objects.py:1301]: Is this actually `null_scalar_products`?

File: pyvista/utilities/geometric_objects.py [hunk 4]
```diff
@@ -1264,9 +1268,82 @@ def Rectangle(points=None):
...
+    if len(non_null_scalar_products) == 0:
+        raise ValueError("The three points should defined orthogonal vectors")
+    if len(non_null_scalar_products) > 1:
+        raise ValueError("Unable to build a rectangle with less than three different points")
+
+    points = np.array([point_0, point_1, point_2, point_0])
+    if np.isclose(scalar_pdct_01_02, 0):
+        points[3] = point_0 + vec_01 + vec_02
+        cells = np.array([[4, 0, 1, 3, 2]])
+    if np.isclose(scalar_pdct_01_12, 0):
```
Comment (MatthewFlamm) [inline pyvista/utilities/geometric_objects.py:1315]: You've already checked above that it has exactly one such case.
```suggestion
    elif np.isclose(scalar_pdct_01_12, 0):
```

File: pyvista/utilities/geometric_objects.py [hunk 5]
```diff
@@ -1264,9 +1268,82 @@ def Rectangle(points=None):
...
+        points[3] = point_1 + vec_12 - vec_01
+        cells = np.array([[4, 0, 1, 2, 3]])
+    if np.isclose(scalar_pdct_02_12, 0):
```
Comment (MatthewFlamm) [inline pyvista/utilities/geometric_objects.py:1318]:
```suggestion
    else:
```

**Review by MatthewFlamm — COMMENTED** [review]
Excellent first PR! Thanks. I have not yet looked through the tests, but some comments to get started.

File: pyvista/utilities/geometric_objects.py (hunk 3 above)
Comment (hippo91) [reply pyvista/utilities/geometric_objects.py:1301]: You are totally right.

File: pyvista/utilities/geometric_objects.py [hunk 6]
```diff
@@ -1264,9 +1268,82 @@ def Rectangle(points=None):
...
+    if len(points) == 4:
+        warnings.warn(
+            'Rectangle defined by 4 points is deprecated. Please use ``pyvista.Quadrilateral``.',
+            PyVistaDeprecationWarning,
+        )
+        return Quadrilateral(points)
+    elif len(points) != 3:
+        raise TypeError('Points must be given as length 3 np.ndarray or list')
+
+    check_valid_vector(points[0], 'points[0]')
```
[the remark this answers is not in the document]
Comment (hippo91) [reply pyvista/utilities/geometric_objects.py]: I changed the code to use your proposal. I did it also in Quadrilateral function and added a test on this last one.

File: pyvista/utilities/geometric_objects.py (hunk 2 above)
Comment (MatthewFlamm) [reply pyvista/utilities/geometric_objects.py]: See discussion in #4146.  I'd recommend reverting back to the original state of hte PR and any cleanup could be done in #4146 if we decide to go in a different direction.

**Review by akaszynski — APPROVED** [review]
I made a few changes, including aligning it with our way of running tests and checking for warnings/errors.

Thanks for your work!

**Comment by hippo91** [comment]
@tkoyama010 @akaszynski @adeak @MatthewFlamm thanks for your review.
Is there something more i can do with this PR?

**Comment by akaszynski** [comment]
> @tkoyama010 @akaszynski @adeak @MatthewFlamm thanks for your review. Is there something more i can do with this PR?

We generally wait 24 hours to allow all contributors to say their peace before merging. Expect it merged tomorrow.

Next release is 0.39.0, probably by the end of this month or sooner.

**Review by tkoyama010 — APPROVED** [review]

**Comment by tkoyama010** [comment]
Great first contribution. Thank you!

File: tests/test_geometric_objects.py [hunk 7]
```diff
@@ -254,19 +256,96 @@ def test_triangle():
...
+
+    # Test all possible orders of the points
+    for pt_tuples in permutations(rotated, 4):
+        mesh = pyvista.Rectangle(list(pt_tuples[0:3]))
+        assert mesh.n_points
+        assert mesh.n_cells
+        assert np.allclose(mesh.points, pt_tuples)
+
+
+def test_rectangle_not_orthognal_entries():
```
Comment (akaszynski) [inline tests/test_geometric_objects.py:309]: Consider adding a docstring for the `test_rectangle_not_orthognal_entries` function to provide more context and describe its purpose.

```suggestion
def test_rectangle_not_orthogonal_entries():
    """Test the error handling when the input points are not orthogonal."""
```

Comment (akaszynski) [reply tests/test_geometric_objects.py:309]: (this is here for testing of another library)

**Pull request merged**